File: livingrimoire start here/LivinGrimoire swift/livingrimoire packet/AlgParts.swift
//
//  AlgParts.swift
//  LivinGrimoireSwiftV1
//
//  Created by moti barski on 09/08/2025.
//

import Foundation


class APSleep: AlgPart {
    var wakeners: Responder
    var done: Bool = false
    var timeGate: TimeGate

    init(wakeners: Responder, sleepMinutes: Int) {
        self.wakeners = wakeners
        self.timeGate = TimeGate(sleepMinutes)
        super.init()
        self.timeGate.openForPauseMinutes()
    }

    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        if wakeners.responsesContainsStr(ear) || timeGate.isClosed() {
            done = true
            return "i am awake"
        }
        if !ear.isEmpty {
            return "zzz"
        }
        return ""
    }

    override func completed() -> Bool {
        return done
    }
}


class APsay:AlgPart{
    var at:Int=10
    var param:String="hmm"
    convenience init(repetitions:Int, param:String) {
        self.init()
        if repetitions<at {self.at=repetitions}
        self.param=param
    }
    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        var axnStr=""
        if self.at>0{
            if ear.lowercased() != self.param{
                axnStr=self.param
                self.at -= 1
            }
        }
        return axnStr
    }
    override func completed() -> Bool {
        return self.at < 1
    }
}


class APMad: AlgPart {
    private var sentences: [String]

    init(_ sentences: String...) {
        // Handle both variadic strings and single array case
        if sentences.count == 1, let single = sentences.first, single.hasPrefix("[") {
            let cleaned = single.replacingOccurrences(of: "[", with: "")
                               .replacingOccurrences(of: "]", with: "")
                               .replacingOccurrences(of: " ", with: "")
            self.sentences = cleaned.components(separatedBy: ",")
        } else {
            self.sentences = sentences
        }
        super.init()
    }

    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        return sentences.isEmpty ? "" : sentences.removeFirst()
    }

    override func completed() -> Bool {
        return sentences.isEmpty
    }
}


class APShy: AlgPart {
    private var sentences: [String] = []

    init(_ sentences: String...) {
        super.init()
        
        if sentences.count == 1 && sentences[0].starts(with: "[") {
            let listString = sentences[0]
                .replacingOccurrences(of: "[", with: "")
                .replacingOccurrences(of: "]", with: "")
                .replacingOccurrences(of: " ", with: "")
            self.sentences = listString.components(separatedBy: ",")
        } else {
            self.sentences = sentences
        }
    }

    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        guard !sentences.isEmpty else { return "" }
        return sentences.removeFirst()
    }

    override func completed() -> Bool {
        return sentences.isEmpty
    }
}


class APHappy: AlgPart {
    private var sentences: [String]

    init(_ sentences: String...) {
        // Handle both variadic strings and single array case
        if sentences.count == 1 && sentences[0].hasPrefix("[") {
            let cleaned = sentences[0]
                .replacingOccurrences(of: "[", with: "")
                .replacingOccurrences(of: "]", with: "")
                .replacingOccurrences(of: " ", with: "")
            self.sentences = cleaned.components(separatedBy: ",")
        } else {
            self.sentences = sentences
        }
        super.init()
    }

    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        return sentences.isEmpty ? "" : sentences.removeFirst()
    }

    override func completed() -> Bool {
        return sentences.isEmpty
    }
}


class APSad: AlgPart {
    private var sentences: [String] = []
    
    init(_ sentences: String...) {
        if sentences.count == 1 && sentences[0].starts(with: "[") {
            // Handle list case
            let cleanString = sentences[0]
                .replacingOccurrences(of: "[", with: "")
                .replacingOccurrences(of: "]", with: "")
                .trimmingCharacters(in: .whitespaces)
            self.sentences = cleanString.components(separatedBy: ",")
        } else {
            // Handle normal case
            self.sentences = sentences
        }
        super.init()
    }

    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        guard !sentences.isEmpty else { return "" }
        return sentences.removeFirst()
    }

    override func completed() -> Bool {
        return sentences.isEmpty
    }
}

File: livingrimoire start here/LivinGrimoire C_sharp/AlgParts.cs
using System.Collections.Generic;
using System.Linq;

public class APSleep : AlgPart
{
    private Responder wakeners;
    private bool done = false;
    private TimeGate timeGate;

    public APSleep(Responder wakeners, int sleepMinutes)
    {
        this.wakeners = wakeners;
        this.timeGate = new TimeGate(sleepMinutes);
        this.timeGate.OpenGate();
    }

    public override string Action(string ear, string skin, string eye)
    {
        if (wakeners.ResponsesContainsStr(ear) || timeGate.IsClosed())
        {
            done = true;
            return "i am awake";
        }
        if (!string.IsNullOrEmpty(ear))
        {
            return "zzz";
        }
        return "";
    }

    public override bool Completed()
    {
        return done;
    }
}

public class APsay : AlgPart
{
    private int at = 10;
    private string param = "hmm";

    public APsay() { }

    public APsay(int repetitions, string param)
    {
        if (repetitions < at) { this.at = repetitions; }
        this.param = param;
    }

    public override string Action(string ear, string skin, string eye)
    {
        string axnStr = "";
        if (this.at > 0)
        {
            if (ear.ToLower() != this.param)
            {
                axnStr = this.param;
                this.at -= 1;
            }
        }
        return axnStr;
    }

    public override bool Completed()
    {
        return this.at < 1;
    }
}

public class APMad : AlgPart
{
    private readonly Queue<string> _sentences;

    // Constructor for params
    public APMad(params string[] sentences)
    {
        _sentences = new Queue<string>(sentences);
    }

    // Constructor for List
    public APMad(List<string> list1)
    {
        _sentences = new Queue<string>(list1);
    }

    // O(1) dequeue
    public override string Action(string ear, string skin, string eye)
    {
        return _sentences.Count > 0 ? _sentences.Dequeue() : "";
    }

    // O(1) completed check
    public override bool Completed()
    {
        return _sentences.Count == 0;
    }
}

public class APShy : AlgPart
{
    private readonly Queue<string> _sentences;

    // Constructor for params
    public APShy(params string[] sentences)
    {
        _sentences = new Queue<string>(sentences);
    }

    // Constructor for List
    public APShy(List<string> list1)
    {
        _sentences = new Queue<string>(list1);
    }

    // O(1) dequeue
    public override string Action(string ear, string skin, string eye)
    {
        return _sentences.Count > 0 ? _sentences.Dequeue() : "";
    }

    // O(1) completed check
    public override bool Completed()
    {
        return _sentences.Count == 0;
    }
}

public class APHappy : AlgPart
{
    private readonly Queue<string> _sentences;

    // Constructor for params
    public APHappy(params string[] sentences)
    {
        _sentences = new Queue<string>(sentences);
    }

    // Constructor for List
    public APHappy(List<string> list1)
    {
        _sentences = new Queue<string>(list1);
    }

    // O(1) dequeue
    public override string Action(string ear, string skin, string eye)
    {
        return _sentences.Count > 0 ? _sentences.Dequeue() : "";
    }

    // O(1) completed check
    public override bool Completed()
    {
        return _sentences.Count == 0;
    }
}

public class APSad : AlgPart
{
    private readonly Queue<string> _sentences;

    // Constructor for params
    public APSad(params string[] sentences)
    {
        _sentences = new Queue<string>(sentences);
    }

    // Constructor for List
    public APSad(List<string> list1)
    {
        _sentences = new Queue<string>(list1);
    }

    // O(1) dequeue
    public override string Action(string ear, string skin, string eye)
    {
        return _sentences.Count > 0 ? _sentences.Dequeue() : "";
    }

    // O(1) completed check
    public override bool Completed()
    {
        return _sentences.Count == 0;
    }
}